### ISSUE TITLE
Should we support turning off pre-rotation and if so, how

### ISSUE THREAD

**Issue by swcurran** [decentralized-identity/didwebvh#151]
Discussed at the Work Item [Meeting of 2024-12-05](https://hackmd.io/k4cIK9vQSlaeg2pdHE51IQ?view#Meeting---05-Dec-2024).

Do we want to allow turning off pre-rotation once started?

We probably do want to allow prerotation to be turned off.  If so:

- Mechanism is to not include the `nextKeyHashes` array in the parameters.
- In that log entry, the `updateKeys` array **MUST** contain keys committed to in the prior `nextKeyHashes` (e.g., pre-rotation active rules apply).
- In the next log entry, the "no pre-rotation rules" apply, and the auth proof must use a key from the `updateKeys` from the previous entry (e.g. same keys as signed the previous entry).

**Comment by swcurran** [comment]
@bj-ms -- what do you think?

**Comment by bj-ms** [comment]
Makes it a bit more complex to handle those state transition but it gives more flexibility at time where they would just create a new did because they don't want to use pre-rotation anymore.

I would agree to that proposition, we just have to be cautious with the implementations and the test-suites to get it right.

**Comment by brianorwhatever** [comment]
it makes me feel uneasy but I haven't come up with any objections yet. We are assuming if a DID's next keys get compromised the malicious actor can already takeover control right? It feels odd to be degrading security levels I think is where my uneasiness comes from.

**Comment by swcurran** [comment]
I think if we are defending against a malicious actor that has the next keys (and web server access), they would probably keep the pre-rotation as it prevents the party “fixing” the log — they wouldn’t have the new “next keys”.

@PatStLouis — this is a reminder that a web server implementation needs to be careful about what it is using for authorization to publish a new log.  We would like to make the access to the publication mechanism different from the authorization to update the log.  Outside of the spec, but something to think about.

**Comment by brianorwhatever** [comment]
![did_webh prerotation (1)](https://github.com/user-attachments/assets/db050f4e-509c-42d7-ac01-9d650fbed3b7)
Is this what disabling prerotation would look like? Note the key `c` is used for 2 proofs in a row. Is that an issue? It means if you want to rotate keys as well as change your update keys it would take 2 log entries.

**Comment by swcurran** [comment]
Yes — thanks — we can use that diagram as well :-)

**Comment by swcurran** [comment]
I don’t think that is an issue.  There are two things that a DID Controller can do:

- Accept that C is used twice in a row in this case — and in other cases.  When not using pre-rotation, that is completely acceptable.  You set `updateKeys` in a version and don’t include it in the next entry(ies), signing that entry and the ones after that with the same key — until you do a version with a new `updateKeys` value.
- If it is a concern, create two entries in a row, and don’t publish the DID Log until the second entry is complete.  Quite a lot of complexity for a weird edge case, but up to the DID Controller.
  - That said, as noted in the implementation guide — I think it is possible that an implementation might choose to do two entries when using pre-rotation, keeping the two key management systems entirely separate — one purely for longer-lived rotations, one for longer-lived use.

**Comment by brianorwhatever** [comment]
ok so this came up for me when I was looking at deactivate. Just talking this through here..
The test is effectively deactivating prerotation (because nextKeyHashes is empty). This doesn't take effect until the next version (which won't exist in a deactivation scenario). However it also doesn't have any updateKeys so it is failing verification there as it doesn't know what the verificationMethod for the proof is. So I think I might be hitting and edge case..

If we have prerotation enabled deactivation log entries must still have updateKeys

**Comment by swcurran** [comment]
So add a requirement for that — mentioning deactivation?

If the DID Controller is turning off pre-rotation (providing an empty `nextUpdateHashes` list) as part of deactivating the DID, the `updateKeys` **MUST** follow the peroration rules.

**Comment by brianorwhatever** [comment]
yes but unfortunately the spec also currently says

> A [DID](https://identity.foundation/didwebvh/next/#term:did) SHOULD update the [DIDDoc](https://identity.foundation/didwebvh/next/#term:diddoc) and parameters object to further indicate the deactivation of the DID, such as including an empty updateKeys list ("updateKeys": []) in the [parameters](https://identity.foundation/didwebvh/next/#term:parameters), preventing further versions of the DID.

So I'm not sure what to do.. Since having `updateKeys` and no `nextKeyHashes` means they could  technically update the DID with those update keys IF the `deactivated` parameter is ignored we are effectively relying entirely on that parameter. Is that a bad thing? Probably not just something to think about

**Comment by brianorwhatever** [comment]
I think we should probably leave it as is and have the special case here be determined by that parameter. Something like "If the DID is being deactivated while prerotation is enabled `updateKeys` should be empty and the `nextKeyHashes` validation is ignored.

**Comment by brianorwhatever** [comment]
erm nevermind.. i guess we need to keep the `updateKeys` to have a verification method for the proof.. so what you said above is more right but like I said relies solely on the `deactivated` parameter

**Comment by swcurran** [comment]
I've tweaked the wording of deactivate to say:

`If the DID is using [[ref: pre-rotation]] two [[ref: DID log entries]] are required, the first to stop the use of pre-rotation, and the second to include the empty `updateKeys` list. For additional details see the section of this specification on [pre-rotation](#pre-rotation-key-hash-generation-and-verification).`

It is a **SHOULD** on the emptying of the `updateKeys`.  The important thing is the `"deactivated": true`.

**Comment by swcurran** [comment]
Spec has been updated and v0.5 reflects what was defined here.

Closing.
